mmtk/src/julia_scanning.rs

```diff
@@ -96,6 +107,63 @@ pub unsafe fn scan_julia_object<EV: EdgeVisitor<JuliaVMEdge>>(obj: Address, clos
         } else if flags.how_custom() == 3 {
             // has a pointer to the object that owns the data
             let owner_addr = mmtk_jl_array_data_owner_addr(array);
+
+            if mmtk_object_is_managed_by_mmtk((*array).data as usize) {
+                let data_owner = owner_addr.load::<ObjectReference>();
+                // if the owner moves and a->data points to it, it needs to be updated as well
+                // first check if it has moved already (as we need to query type info from it)
+                let owner =
+                    if mmtk_object_is_managed_by_mmtk(data_owner.to_raw_address().as_usize()) {
+                        mmtk_get_possibly_forwared(data_owner)
+                    } else {
+                        data_owner
+                    };
+
+                // owner may be either a string or another array
```

Comment:
This code looks adhoc and hacky. I guess it could be a pain to maintain. If internal reference is a rare case, you may consider pinning the object so it won't move.

Reply:
Fixed.